core/src/main/scala/kafka/server/metadata/BrokerMetadataEventManager.scala

```diff
@@ -0,0 +1,200 @@
+ *
+ * The manager injects this event into the event stream such that processors receive it as soon as they finish their
+ * current message batch, if any, otherwise they receive it immediately.
+ *
+ * @param brokerEpoch the broker epoch that was fenced
+ */
+final case class OutOfBandFenceLocalBrokerEvent(brokerEpoch: Long) extends BrokerMetadataEvent
+
+/*
+ * WakeupEvent is necessary for the case when the event thread is blocked in queue.take() and we need to wake it up.
+ * This event has no other semantic meaning and should be ignored when seen.  It is useful in two specific situations:
+ * 1) when an out-of-band message is available; and 2) when shutdown needs to occur.  The presence of this message
+ * ensures these other messages are seen quickly when no other messages are in the queue.
+ */
+final case object WakeupEvent extends BrokerMetadataEvent
+
+class QueuedEvent(val event: BrokerMetadataEvent, val enqueueTimeMs: Long) {
+  override def toString: String = {
+    s"QueuedEvent(event=$event, enqueueTimeMs=$enqueueTimeMs)"
+  }
+}
+
+trait MetadataEventProcessor {
+  def currentMetadataOffset(): Long
+  def processStartup(): Unit
+  def process(apiMessages: List[ApiMessage], lastMetadataOffset: Long): Unit
+  def process(registerLocalBrokerEvent: OutOfBandRegisterLocalBrokerEvent): Unit
+  def process(fenceLocalBrokerEvent: OutOfBandFenceLocalBrokerEvent): Unit
+}
+
+class BrokerMetadataEventManager(config: KafkaConfig,
+                                 time: Time,
+                                 processor: MetadataEventProcessor,
+                                 eventQueueTimeHistogramTimeoutMs: Long = 300000) extends KafkaMetricsGroup {
+
+  private val blockingQueue = new LinkedBlockingQueue[QueuedEvent]
+  private val bufferQueue: util.Queue[QueuedEvent] = new util.ArrayDeque[QueuedEvent]
+  private val outOfBandQueue = new ConcurrentLinkedQueue[QueuedEvent]
+  @volatile private var totalQueueSize: Int = 0
+
+  private val thread = new BrokerMetadataEventThread(
+    s"${BrokerMetadataEventManager.BrokerMetadataEventThreadNamePrefix}${config.brokerId}${BrokerMetadataEventManager.BrokerMetadataEventThreadNameSuffix}")
+
+  // metrics
+  private val eventQueueTimeHist = newHistogram(BrokerMetadataEventManager.EventQueueTimeMetricName)
+  newGauge(BrokerMetadataEventManager.EventQueueSizeMetricName, () => totalQueueSize)
+
+  def start(): Unit = {
+    put(StartupEvent)
+    thread.start()
+  }
+
+  def close(): Unit = {
+    try {
+      thread.initiateShutdown()
+      put(WakeupEvent) // wake up the thread in case it is blocked on queue.take()
+      thread.awaitShutdown()
+    } finally {
+      removeMetric(BrokerMetadataEventManager.EventQueueTimeMetricName)
+      removeMetric(BrokerMetadataEventManager.EventQueueSizeMetricName)
+    }
+  }
+
+  def put(event: BrokerMetadataEvent): QueuedEvent = {
+    val queuedEvent = new QueuedEvent(event, time.milliseconds())
```

Comment:
I realize that not all of our code follows this convention, but interval timing is really supposed to use `time.nanoseconds()` rather than `time.milliseconds()`.  The reason is because `time.nanoseconds()` uses a monotonic timer, which is unaffected by clock changes, whereas `time.milliseconds` tracks the wall-clock time.  So with `time.milliseconds` the user could set the system clock one hour forward and then Kafka would think that whatever event was ongoing took a full hour, for example.

For some reason, the JVM has always exposed monotonic times with nanosecond granularity and wall-clock times with millisecond granularity, which has trained everyone to think that nanoseconds == monotonic.  Of course the underlying the operating system also offers wall-clock time with nanosecond granularity, but they never exposed that, for whatever reason.  Maybe there will be a new JDK revision some day that does?